### ISSUE TITLE
CKEditor 5 integrations for JavaScript frameworks (React, Angular 2+, Vue.js, ??)

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5#1002]
Soon (in May 2018, in fact) we'll be starting to work on integrations for the most popular JS frameworks. Most likely, we'll be working on them one by one and we'll start from React.

We don't know yet any more details, but I wanted to create this ticket so we can start gathering feedback... and so I can link to somewhere from the upcoming blog post ;).

TODO: 

* [x] React integration (ticket: https://github.com/ckeditor/ckeditor5/issues/1015) – https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/frameworks/react.html  ✅ 
* [x] Angular integration (ticket: https://github.com/ckeditor/ckeditor5/issues/1020) – https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/frameworks/angular.html  ✅
* [x] Vue.js integration (ticket: https://github.com/ckeditor/ckeditor5-vue/issues/1) – https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/frameworks/vuejs.html ✅
* [ ] What should we work on next? *(you can propose a framework in a comment or 👍  an existing comment)*

**Comment by Reinmar** [comment]
Two related topics:

* https://github.com/ckeditor/ckeditor5/issues/599 – [Docs] Example integration with React, Angular 1/2+, Vue, Ember
* https://github.com/ckeditor/ckeditor5/issues/347 – Angular 2 integration - WYSIWYG Editor #FeatureRequest

**Comment by cavemansspa** [comment]
Here's a very simple Mithril / CKEditor [integration](https://flems.io/#0=N4IgtglgJlA2CmIBcBWAzAOgEwHYA0IAxgPYB2AzsQskVbAIYAO58UIBAZhAucgNqhS9MIiQgMACwAuYWO1qkp8RTQA85QgCcIjKQAJymwgF4AOiGlTmSAPQ3CUUhkIBrVhCnFNz4mHtuoDy8UGwBGAAYMSPD7BnJyCEJ-d09vACtycwA+VRsNbV0s+RYEQikIMl4xcKQADhQQAF88QWFRcQz5EkVlKRpTUm7yfQBhFwBRQNS9Yz0OAFdB8rIACgA3UmIoeABKPWABvT1NeCl5zVJ9w6O9NYh4AHckPXXN7b3jLKvLm6OTs4uejAKwA5M4AkFNCCdtcjs1YXoEWQtPB6Epnq8trsZl8Dj9fiM4glCJNIQNTFIKVSpM4Tmj4Ctbm94BgoL49DDSNTqRgpBJlIyUl4cd9uZTxUMqCzYMQAOaCqbCnYAbgRvyOGyxGDRUk05G1jEYAGUpPTwaTprMhZpyeLqY0Obaxc40YQJILNJphZ9RXa-ZKEBh4J6vB6vZoOaquX6KQ6VQjGtdE6RkwMhvoAIKGmZzRZlCqkTHvb5Ha4IfRMY2mpQ5vG-VwWrzPRbbLikVh4BHSWTN0itiDtqBJqOln7-c6XOs3O6PDGa4s+4EgkF4PR8NVA0ESUIrvQggASqMCpFl0M7+JuwLGjc0q+AlZN9OeD+r8EaO3P6s3IIARvMpJ4pC7lOX7IrAiQuBi2I+iBX56AG0pyisL70pycE3ChSiSDIsA5phLINoq3iyqcAAiaL0CsaFwcmX7NHuhIQWeG5LruYC8po8zDMhhqPlh3awDs1FHAAutRyapqQ7FgMQixSCsbKEPMIiKBgP5bAAnquWaMDsxTwKUywUDQWAoEgoQAJxNC0IBCCINDOPEXRkEoKhiE0IkEOBpAuFUAi2W0NCQHy2hyAQ5xyGIljWHYiyMC4so+H4wUSKFAACoQYJlABsNgpaFGCdAQUgaYw7T5DofTNK09liD+TnhZokUWABMU2HFCVJTY9XkBlWUYCEPWFbwxWleVWiVR5jRAA).

Do you have some ideas on what you'd like to see as framework integration points?

**Comment by Reinmar** [comment]
Thanks for sharing!

In fact, @szymonkups created some time ago a quick integration with React too: https://jsfiddle.net/szymonkups/43hhtk0j/.

**Comment by cavemansspa** [comment]
Here's a more comprehensive [example](https://flems.io/#0=N4IgtglgJlA2CmIBcBmAjAOgKwDYA0IAZhAgM7IDaoAdgIZiJIgYAWALmLCAQMYD21NvEHIQAHhbxaUAHwAdamNgRqAawAEAJ3iwAvHJCk2ATzKT4bA+pbbC+kOzYAHUkgD0bnlGoArUhh5YPgBXKEJYWm0AvjA3Wh9aAA83ZQAjUjdSeDBaQQgeAFpgiDcAJgwUDDRM7Ny2fIxIagDSUisVIQBzTQgTe1IWWlLcAr4AEQBhAAVSqAA5AHU0AHYATTQARRYAIQB3DYBqYNLVABlaAGlEvlpghYA3AEkAcVVSFFoAFlpSAE4Ni6kezqHiaPitPg9Toqey5ATGMAhNogdRuGTqdQKMRuSTSeSKUigiBONjqUiaHj2RwudyebwBVTwKC9SHRWI8RnMtiQrBuNAABgwgv5ngirXynk5LM0GD8Bhk2MJPRJMm4hh08B49QE5CYKCQAA4sCAAL54Gj0RggCaqACiXNZfjV-EEwjYohdRnUNvt0vUunUhGC1C1EAEAAp7tQ+FB4ABKdTABQY5PqBCkoy0IT+xOpjFZBBayFIdRRmPwDBZtiafwFzXczTqAA+TfUAHIGUzpW28Hn1F2GxMBMROiWy7HK2xq-4B5Ch9QR83W8AzX2Ob6GyXg7HiNQmXhUW403weLRYFp4IQ++xOGPoxOqzXWBxzy31NvLyomSnqBiTanU20Nhgk0X8k1-DFMQgjF7ggeBdhLSN73jf10XAyDIKAkDfzAcNMyEDA6yLTQ4z7VdoKgjCBFBKQhEQ8cUN0NC+wxCYxVIfIN0hBQ5EsPjeICbQs3gcNS2QjAoBiA98IrWdNHnRdSOoXiVL4jA2EkahRLk1DcwojD8zYYSGS4xsAzkliDIgQgkPLSdpwwAQJiEoQAGVgh4Hh4FaBMGPsp8nJc+B3M87z-FPWBYHDahgkig85KU1TVJNdREv4tTTzYHgWG0zQwTM5jlPS1TPT4BAMHgPLIVy-LUoAbh44q2BSuMGoo8iML7ARUkvSFKuyPh7ngejkITJi9IMskjII9cHRlWMjDBYxwyUjD-2UiD1pNNUiO1ahdRAUpSiQI7TXNEA6AYUQJloIacn21ypgAQVlcheAEIQRCYT1SRuu7clIR6npzIMQz29RbNjBNwL7dMpuEnN0MmkE7TmrdqB3L8oF7fTIJvWB0cxvdscsjEIFIDcVFHdR+Rx5H1CgLNaBLJH6YxJxwTYVx1AoUm2eAaASzQaTgjAHJNGMEs21IUXxeMdQ0B7awXylsQnBkDTyfULXSBieBlc4BXsXVpXgiyTQpaMeBYPgNsOrZyaBagEtShFsXIkl9sZfdiX1FKJX8dV9XNdIbXQ91hgDfPUpjZkU3zaloawWoCBVFoO26YdgynZLFA3bly3ZY99QUADlX2zVjWWC1nW9ajkvY-jyqpZ8YRqGMDO+YwgBdTOHbNyrudZrOMVb6h25Z4gazYOZLSlgApNuO4PMUZ7n9sAAkws73GHaTgRU+ZxMp6MWeGClgA1SqD7TpXV7P4b2wAeXFHeR4wq2bcniBp4fqXXKEDbO+Pw17n3bAAMR0PkCAcI7Zd0gutemiCEFtT7FhUCE1kY2wQhDBiY1CrwIxOgnC4YOzFAwMoNQGBejZFIErXmu82YyQkkzDAHMjD+ByE4SG8BHhCDAKtLO3dEqMIwqDUMAgxLlj4dkcM7C2DQ0IQZOGGl4CmRzHIzm1CoD+l0AGZh5NKbUE6AmAA-FI2MplNFGATITT8xMlEYThgPMy8MCKMyMhgFxpAKDyK8ebbujjIJwwgGATorkKQ5jbNSVwHgsj3XqIUch-BYihNoJ0bycR7hM00CkSIGS3BtnUAcdQfiXHFPbLKJwnQ2xBIxHDfGOY-GNPMWAdSmgzZsGsWwZ8nBbHtltK6RsEd9axiMiQfwGAMC1I2u-TCFhsI8zqQZXCHYaFgCVsPOZDtqLKA5CNcs+C3EVkMVyKmTStHQGWUgg8DDtnv1WV4iA6kVDGGoTkDJStVmhM6BQcklIiklJ+REngFS2zdzbHGOMfd7ls1wgYaIrpBAGFudc7Z8LmC4ljJoFF75zYYBPqA-WJSimArxZVChICH7QrRXMjFElvJEhJGGZS3Ao6CNhWzYRtKDKBNEVnGl-KHaqNMssvlI91pKPEeDBiVj5GKKFQZYh6hHnkKyJ0Bggh6E8twj6OaB4tmcqjiWZpL4YVGvUIFWiwUPJeVaIhBKulDUWrJoQCGpq+n9jmoRCwYwmbdN6bADlLqP7TQrLNP05k5o8rWjyk0gqdWkKeeoVIwQpw6m1YqkeqrnndHgMIDAqb03UE2TG5GRhTCP2AOLaE1AAAqfAnBSzQNkN+IaDK7PyKoB1RznXtsMsZU5vRzkBn5GWh2Hrzz6LDSZb1GS2B+qMitcdBlkEhrNO2Vyt1baCv7TmjA2goCFrTdyEtBqV2hsrSzGtKgG1NvbC2jZ9t+0Yk7fsiGjFCovvLTOod9RjE5jHVmo1zDGnTuMhGhsGB52LtoMu4Dcy10uo3W2G6IYdCQpxjxBDvKRH3O5YqpDsb2rYe2gQXaLKDooCwEgT4BozoWiukwV6zoPpulEG0xEwYumSR4KLN0haYzGAPH9WoD1npxgYxdS0ogWhvRAC6T67omCml7iAShbxKCMatJADSPQuAEBAlwJgMTaTBicKoTobI3C6ZsCQAAApgTAOAbO9Ds7AFjBATBOCtEqYk7ozTadEOkeTRnRCmY8OZyz1mQuOaqNgNwIXPMgG875plAXu4miAA).

**Comment by pomek** [comment]
The first version of React integration has just been published. 

Now using `npm install @ckeditor/ckeditor5-react` will install the package in version `1.0.0-beta.1`.

The feedback is welcome.

**Comment by mmonirul** [comment]
Is there any vue typescript integration?

**Comment by ma2ciek** [comment]
There's no vue integration at this moment.

If you know well this library you can create an implementation of the integration and share it in the https://github.com/ckeditor/ckeditor5/issues/599 thread. I've found a little bit outdated one there - https://github.com/ckeditor/ckeditor5/issues/599#issuecomment-345202341 - the `changesDone` event must be replaced with `change:data` and after that change, the code should work.

**Comment by Reinmar** [comment]
We're starting work on the Vue.js component very soon. You can track the progress in https://github.com/ckeditor/ckeditor5-vue/issues/1.

**Comment by tajchumber** [comment]
how about integrating Elm with CKEditor?

**Comment by Reinmar** [comment]
Interesting. What such integration would need to do? AFAIR Elm is a language which compiles to JS, so I'd guess you can use JS APIs from Elm's level?

**Comment by tajchumber** [comment]
yes it does compile to Javascript.

The way Elm "comunicates" with normal javscript is through the use of [Ports](https://guide.elm-lang.org/interop/ports.html).

It would be nice if CKEditor to already have some ports by default ready for use.
For example to get the return value `editor.getData()` or to create an editor based on an event managed by Elm.

**Comment by Reinmar** [comment]
The first version of Vue.js integration was released on npm: https://www.npmjs.com/package/@ckeditor/ckeditor5-vue 🎉 🎉 🎉 

You can find the documentation in https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/frameworks/vuejs.html

**Comment by sauravraj95** [comment]
it will be great if we get the access to the text area so that we can dynamically paste some emoji or any icon while copying it to clip board thing in angular. Or simple have some features like this.

**Comment by Reinmar** [comment]
@sauravraj95, please report a separate issue for this in https://github.com/ckeditor/ckeditor5-angular.

**Comment by Reinmar** [comment]
OK, we've introduced the main 3 integrations and people open separate tickets for othe libs/cmses so we can close this ticket now.

**Comment by josegetout** [comment]
> Just FYI. We have this working in ember fine, but internally. Waiting on word paste before we can completely replace our current editor. Then will see about releasing publically.

Hi! I'm struggling to integrate ckeditor5 with Ember, how did you make it? Any tip for this Ember newbie :pray:

**Comment by robclancy** [comment]
Published a new version with more details (but still lacking). https://github.com/postedin/ember-ckeditor

**Comment by jcmiron-ps** [comment]
Any plans for Blazor?

**Comment by xinlingfeiwu** [comment]
**When can there be an official example of integration with Extjs7 ？**

**Comment by akshay-solutelabs** [comment]
Is there any reference for the rails 6 with CKEditor 5?

**Comment by xinlingfeiwu** [comment]
您好，邮件已经收到，我会尽快给您回复。

**Comment by tbreuss** [comment]
> Here's a more comprehensive [example](https://flems.io/#0=N4IgtglgJlA2CmIBcBmAjAOgKwDYA0IAZhAgM7IDaoAdgIZiJIgYAWALmLCAQMYD21NvEHIQAHhbxaUAHwAdamNgRqAawAEAJ3iwAvHJCk2ATzKT4bA+pbbC+kOzYAHUkgD0bnlGoArUhh5YPgBXKEJYWm0AvjA3Wh9aAA83ZQAjUjdSeDBaQQgeAFpgiDcAJgwUDDRM7Ny2fIxIagDSUisVIQBzTQgTe1IWWlLcAr4AEQBhAAVSqAA5AHU0AHYATTQARRYAIQB3DYBqYNLVABlaAGlEvlpghYA3AEkAcVVSFFoAFlpSAE4Ni6kezqHiaPitPg9Toqey5ATGMAhNogdRuGTqdQKMRuSTSeSKUigiBONjqUiaHj2RwudyebwBVTwKC9SHRWI8RnMtiQrBuNAABgwgv5ngirXynk5LM0GD8Bhk2MJPRJMm4hh08B49QE5CYKCQAA4sCAAL54Gj0RggCaqACiXNZfjV-EEwjYohdRnUNvt0vUunUhGC1C1EAEAAp7tQ+FB4ABKdTABQY5PqBCkoy0IT+xOpjFZBBayFIdRRmPwDBZtiafwFzXczTqAA+TfUAHIGUzpW28Hn1F2GxMBMROiWy7HK2xq-4B5Ch9QR83W8AzX2Ob6GyXg7HiNQmXhUW403weLRYFp4IQ++xOGPoxOqzXWBxzy31NvLyomSnqBiTanU20Nhgk0X8k1-DFMQgjF7ggeBdhLSN73jf10XAyDIKAkDfzAcNMyEDA6yLTQ4z7VdoKgjCBFBKQhEQ8cUN0NC+wxCYxVIfIN0hBQ5EsPjeICbQs3gcNS2QjAoBiA98IrWdNHnRdSOoXiVL4jA2EkahRLk1DcwojD8zYYSGS4xsAzkliDIgQgkPLSdpwwAQJiEoQAGVgh4Hh4FaBMGPsp8nJc+B3M87z-FPWBYHDahgkig85KU1TVJNdREv4tTTzYHgWG0zQwTM5jlPS1TPT4BAMHgPLIVy-LUoAbh44q2BSuMGoo8iML7ARUkvSFKuyPh7ngejkITJi9IMskjII9cHRlWMjDBYxwyUjD-2UiD1pNNUiO1ahdRAUpSiQI7TXNEA6AYUQJloIacn21ypgAQVlcheAEIQRCYT1SRuu7clIR6npzIMQz29RbNjBNwL7dMpuEnN0MmkE7TmrdqB3L8oF7fTIJvWB0cxvdscsjEIFIDcVFHdR+Rx5H1CgLNaBLJH6YxJxwTYVx1AoUm2eAaASzQaTgjAHJNGMEs21IUXxeMdQ0B7awXylsQnBkDTyfULXSBieBlc4BXsXVpXgiyTQpaMeBYPgNsOrZyaBagEtShFsXIkl9sZfdiX1FKJX8dV9XNdIbXQ91hgDfPUpjZkU3zaloawWoCBVFoO26YdgynZLFA3bly3ZY99QUADlX2zVjWWC1nW9ajkvY-jyqpZ8YRqGMDO+YwgBdTOHbNyrudZrOMVb6h25Z4gazYOZLSlgApNuO4PMUZ7n9sAAkws73GHaTgRU+ZxMp6MWeGClgA1SqD7TpXV7P4b2wAeXFHeR4wq2bcniBp4fqXXKEDbO+Pw17n3bAAMR0PkCAcI7Zd0gutemiCEFtT7FhUCE1kY2wQhDBiY1CrwIxOgnC4YOzFAwMoNQGBejZFIErXmu82YyQkkzDAHMjD+ByE4SG8BHhCDAKtLO3dEqMIwqDUMAgxLlj4dkcM7C2DQ0IQZOGGl4CmRzHIzm1CoD+l0AGZh5NKbUE6AmAA-FI2MplNFGATITT8xMlEYThgPMy8MCKMyMhgFxpAKDyK8ebbujjIJwwgGATorkKQ5jbNSVwHgsj3XqIUch-BYihNoJ0bycR7hM00CkSIGS3BtnUAcdQfiXHFPbLKJwnQ2xBIxHDfGOY-GNPMWAdSmgzZsGsWwZ8nBbHtltK6RsEd9axiMiQfwGAMC1I2u-TCFhsI8zqQZXCHYaFgCVsPOZDtqLKA5CNcs+C3EVkMVyKmTStHQGWUgg8DDtnv1WV4iA6kVDGGoTkDJStVmhM6BQcklIiklJ+REngFS2zdzbHGOMfd7ls1wgYaIrpBAGFudc7Z8LmC4ljJoFF75zYYBPqA-WJSimArxZVChICH7QrRXMjFElvJEhJGGZS3Ao6CNhWzYRtKDKBNEVnGl-KHaqNMssvlI91pKPEeDBiVj5GKKFQZYh6hHnkKyJ0Bggh6E8twj6OaB4tmcqjiWZpL4YVGvUIFWiwUPJeVaIhBKulDUWrJoQCGpq+n9jmoRCwYwmbdN6bADlLqP7TQrLNP05k5o8rWjyk0gqdWkKeeoVIwQpw6m1YqkeqrnndHgMIDAqb03UE2TG5GRhTCP2AOLaE1AAAqfAnBSzQNkN+IaDK7PyKoB1RznXtsMsZU5vRzkBn5GWh2Hrzz6LDSZb1GS2B+qMitcdBlkEhrNO2Vyt1baCv7TmjA2goCFrTdyEtBqV2hsrSzGtKgG1NvbC2jZ9t+0Yk7fsiGjFCovvLTOod9RjE5jHVmo1zDGnTuMhGhsGB52LtoMu4Dcy10uo3W2G6IYdCQpxjxBDvKRH3O5YqpDsb2rYe2gQXaLKDooCwEgT4BozoWiukwV6zoPpulEG0xEwYumSR4KLN0haYzGAPH9WoD1npxgYxdS0ogWhvRAC6T67omCml7iAShbxKCMatJADSPQuAEBAlwJgMTaTBicKoTobI3C6ZsCQAAApgTAOAbO9Ds7AFjBATBOCtEqYk7ozTadEOkeTRnRCmY8OZyz1mQuOaqNgNwIXPMgG875plAXu4miAA).

Thanks @cavemansspa, your Mithril.js example was very helpful!

**Comment by xinlingfeiwu** [comment]
您好，邮件已经收到，我会尽快给您回复。

**Comment by davidparry** [comment]
Flutter for Web would be a big help. Also in our decision to use CKEditor.

**Comment by xinlingfeiwu** [comment]
您好，邮件已经收到，我会尽快给您回复。
